src/main/java/org/spongepowered/common/world/FakePlayer.java

```diff
@@ -0,0 +1,203 @@
+import org.spongepowered.api.block.BlockState;
+import org.spongepowered.api.data.type.HandTypes;
+import org.spongepowered.api.event.SpongeEventFactory;
+import org.spongepowered.api.event.block.InteractBlockEvent;
+import org.spongepowered.api.event.cause.Cause;
+import org.spongepowered.api.event.cause.NamedCause;
+import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.util.Direction;
+import org.spongepowered.api.world.World;
+import org.spongepowered.common.SpongeImpl;
+import org.spongepowered.common.event.tracking.PhaseContext;
+import org.spongepowered.common.event.tracking.phase.plugin.PluginPhase;
+import org.spongepowered.common.interfaces.world.IMixinWorldServer;
+import org.spongepowered.common.registry.provider.DirectionFacingProvider;
+import org.spongepowered.common.util.VecHelper;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * A marker to state that this is a fake player.
+ */
+public interface FakePlayer {
+
+    Controller controller = new Controller(SpongeImpl.getGame().createFakePlayerFactory());
+
+    public class Factory {
+
+        private final Map<UUID, EntityPlayerMP> players = new MapMaker().weakValues().makeMap();
+
+        public EntityPlayerMP getPlayer(WorldServer world, GameProfile profile) {
+            EntityPlayerMP player = players.get(profile.getId());
+            if (player == null) {
+                players.put(profile.getId(), player = new SimulatedPlayer(world, profile));
+            }
+            return player;
+        }
+
+    }
+
+    public static class Controller {
+
+        private final Factory factory;
+
+        Controller(Factory factory) {
+            this.factory = factory;
+        }
+
+        private EntityPlayerMP setUpPlayer(GameProfile profile, World world, int x, int y, int z, ItemStack itemInHand, Cause cause) {
+            EntityPlayerMP player = factory.getPlayer((WorldServer) world, profile);
+            player.world = (WorldServer) world;
+            player.interactionManager.setWorld((WorldServer) world);
+            player.posX = x;
+            player.posY = y;
+            player.posZ = z;
+            player.onGround = true;
+            net.minecraft.item.ItemStack stack = (net.minecraft.item.ItemStack) itemInHand;
+            if (stack == null) {
+                stack = net.minecraft.item.ItemStack.EMPTY;
+            }
+            player.inventory.setItemStack(stack);
+            player.inventory.mainInventory.set(player.inventory.currentItem, stack);
+
+            if (!cause.containsNamed(NamedCause.SOURCE)) {
+                // Cause tracker REQUIRES source?
+                cause = cause.with(NamedCause.of(NamedCause.SOURCE, cause.get(NamedCause.PLAYER_SIMULATED, Object.class).get()));
+            }
+            // Set up cause capture
+            ((IMixinWorldServer) world).getCauseTracker().switchToPhase(PluginPhase.State.FAKEPLAYER, PhaseContext.start()
```

Comment:
My confusion is not in the pushing of the cause (as that is correct) but where you are doing this.

Aren't FakePlayers singletons per world? If so, you would switchToPhase and toss the fakeplayer on the cause stack when the API call occurs.